src/constraint.rs

```diff
@@ -847,3 +869,118 @@ impl ConstraintValidator for ContainerLengthConstraint {
         Ok(())
     }
 }
+
+/// Implements an `byte_length` constraint of Ion Schema
+/// [byte_length]: https://amzn.github.io/ion-schema/docs/spec.html#byte_length
+#[derive(Debug, Clone, PartialEq)]
+pub struct ByteLengthConstraint {
+    length_range: Range,
+}
+
+impl ByteLengthConstraint {
+    pub fn new(length_range: Range) -> Self {
+        Self { length_range }
+    }
+
+    pub fn length(&self) -> &Range {
+        &self.length_range
+    }
+}
+
+impl ConstraintValidator for ByteLengthConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // Check for null value
+        if value.is_null() {
+            return Err(Violation::new(
+                "byte_length",
+                ViolationCode::TypeMismatched,
+                &format!("expected a clob/blob found {:?}", value),
+            ));
+        }
+
+        // get the size of given bytes
+        let size = match value.ion_type() {
+            IonType::Clob | IonType::Blob => value.as_bytes().unwrap().len(),
+            _ => {
+                // return Violation if value is not an clob/blob
+                return Err(Violation::new(
+                    "byte_length",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected a clob/blob but found {}", value.ion_type()),
+                ));
+            }
+        };
```

Comment:
I will still have to keep null check separate from this. Because for `null.*` kind of value would print its Ion type instead of saying its a null. For example,
value: `null.bool`
Violation message: `expected a clob/blob but found bool`

Reply:
What about this?

```suggestion
        // get the size of given bytes
        let size = match value.as_bytes() {
            Some(bytes) => bytes.len(),
            _ => {
                // return Violation if value is not an clob/blob
                return Err(Violation::new(
                    "byte_length",
                    ViolationCode::TypeMismatched,
                    &format!("expected a clob/blob but found {}", if value.is_null() { value } else { value.ion_type() }),
                ));
            }
        };
```

Reply:
Does this sound good? In order to satisfy rust compiler to have same return types for `if-else` statement.
```rust
// get the size of given bytes
let size = match value.as_bytes() {
    Some(bytes) => bytes.len(),
    _ => {
        // return Violation if value is not an clob/blob
        return Err(Violation::new(
            "byte_length",
            ViolationCode::TypeMismatched,
            &format!(
                "expected a clob/blob but found {}",
                if value.is_null() {
                    format!("{:?}", value)
                } else {
                    format!("{}", value.ion_type())
                }
            ),
        ));
    }
};
```

Reply:
Yes, that looks good. I'm surprised though—I thought that since the result of both branches implemented `Display`, they _could_ be the same (abstract) type.

Reply:
> I'm surprised though—I thought that since the result of both branches implemented Display, they could be the same (abstract) type.

This is right in principle. However, in contexts where you're not using generics ("I'll take any `T` that implements `Display`!") the concrete types have to match. You could achieve this by explicitly casting each arm of the `if/else` to a `&dyn Display` so they were literally the same type on the stack.